### ISSUE TITLE
More 'honest' ``Generic`` instance?

### ISSUE THREAD

**Issue by kozross** [expipiplus1/vector-sized#79]
Currently, the ``Generic`` instance for ``Data.Vector.Sized.Vector`` essentially states 'I am a newtype around a ``Data.Vector``':

```haskell
> import qualified Data.Vector.Sized as VS
> import GHC.Generics
> :set -XTypeInType
> :kind! Rep (VS.Vector 3 Bool)
Rep (VS.Vector 3 Bool) :: * -> *
= D1
    ('MetaData
       "Vector"
       "Data.Vector.Generic.Sized.Internal"
       "vector-sized-1.2.0.1-3ec566a97853b37b4d06b4d57461132f10e5653a32dc092cdf363e56c89119e6"
       'True)
    (C1
       ('MetaCons "Vector" 'PrefixI 'False)
       (S1
          ('MetaSel
             'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
          (Rec0 (vector-0.12.0.3:Data.Vector.Vector Bool))))
```

As I was writing ``finitary``, I noticed that this representation is somewhat dishonest, as it reveals nothing about the 'structure' of a fixed-length vector; in addition to other things, it meant that I couldn't derive a ``Finitary`` instance using ``Data.Vector.Sized.Vector``'s ``Generic`` instance, because the underlying type (that is, ``Data.Vector.Vector a``) isn't ``Finitary`` (and nor can it be).

I believe a much more 'honest' ``Generic`` instance for a type-level sized vector would resemble that of a tuple:

```haskell
> :kind! Rep (Bool, Bool, Bool)
Rep (Bool, Bool, Bool) :: * -> *
= D1
    ('MetaData "(,,)" "GHC.Tuple" "ghc-prim" 'False)
    (C1
       ('MetaCons "(,,)" 'PrefixI 'False)
       (S1
          ('MetaSel
             'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
          (Rec0 Bool)
        :*: (S1
               ('MetaSel
                  'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
               (Rec0 Bool)
             :*: S1
                   ('MetaSel
                      'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
                   (Rec0 Bool))))
```

I care about this issue deeply enough that I am willing to write this 'more honest' instance by hand. At the same time, I understand that this isn't necessarily something we might want, nor is it necessarily a good idea for reasons that might not occur to me. Therefore, I wanted to run it past you first, before I fork this and start working on it.

@mstksg, @expipiplus1: Thoughts? Opinions?

**Comment by mstksg** [comment]
Hm, Generics is admittedly a little awkward.  At its most practical level, it's a way for us to access the ADT declaration `data Foo = Bar Int` programatically in Haskell, "pattern matching" on the actual constructors and names of fields.  So in this sense, it is honest.

However, `Vector` is an abstract type, and its "constructor" is not supposed to exist technically as a part of its public API.  We have it hidden behind an unsafe module, but because typeclasses are global, this leaks.  So because `Vector` is an abstract type, offering its newtype-definition as a `Generic` instance is potentially dishonest, since we are trying our best to hide the fact that it "contains" a `Vector`.  Exposing a `Generic` instance essentially allows for *unsafe* construction of improperly sized Vectors...even though we want to promise that all of Vector's public API is safe.

The road taken by other famous abstract data types in Haskell (like [Data.Map](https://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Lazy.html)) is to *not* provide a `Generic` instance.  Not providing a Generic instance is consistent with not providing the "naked" constructors of `Map`: they're semantically meaningless, and would allow you to create unsound Maps.

So, there is the middle ground that is proposed here, where we hand-write a Generic instance that tells us not the syntactical ADT declaration, but as an abstract typeclass-based interface.  That way, the Generic instance respects the actual *abstract* structure of `Vector`.

I am now recalling that we had the exact same debate at mniip/finite-typelits#5 

I think the general conclusion for me is that either position is "morally" tenable.  However, I think the expectations for people using a `Generic` instance is that it would match the ADT syntax.  People usually expect to have access to things like constructor names, field names, etc.

If a hand-written Generic instance can provide these expectations, then I wouldn't be too opposed to it.  However, it's hard to imagine a situation in this specific case where it would fill these expectations.

Also note that people using `Generic` to derive instances would run into some really weird performance aspects if we have an abstract interface, like:

*   `NFData` -- an ideal correct Generic instance respecting the abstraction would require inductively pattern matching on `n` different linked list cons cells, which is much worse than the instance people might expect
*   `ToJSON`/`FromJSON` -- an ideal correct Generic instance would create...a giant record? like { 0 : a, 1 : b, 2 : c }.  Again, different from expectations.

I think in line with the expectations of people using Generic deriving mechanisms, *hiding* `Generic` might be the most morally correct (not a lie, but also not being able to deliver on expectations of metadata if given as a custom "true" abstract-respecting instance).  However, the utility of `Generics` as an auto-deriving instance mechanism for things like `NFData` makes me feel like it's better left in as auto-derived and unsafe than not in at all.

I didn't expect to write this much, but those are my thoughts :)  And the discussion at the issue linked in finite-typelits is also relevant I believe.

**Comment by kozross** [comment]
@mstksg It's certainly an interesting set of points you're making vis a vis my own take. I agree, in principle, with the idea that we want as close a correspondence between the ``Rep`` and the actual type. I also agree that exposing the fact that we're a wrapper around an _un_sized ``Vector`` is not particularly good - it essentially leaks implementation details.

Now, the situation with ``NFData`` and ``*JSON`` is a valid problem, I agree. In my view, ``NFData`` is the worse case, for the simple reason that a faithful translation to JSON of a fixed-size vector isn't an array (because ... well, it's _not_ an array, actually, since its size is 'burned in'). If people expect it to be an array, I would argue that they're wrong: they're throwing away information. The giant record is a much more faithful representation. However, this is beside the point in this case.

My other idea was to provide a ``newtype`` wrapper around sized ``Vector``s in a different package, called ``StructuralVector`` or something. Then, I could hand-roll the ``Generic`` instance the 'morally correct' way, and then simply write coercive wrappers around the API we currently provide to allow people to use it as if it were a ``Vector`` if they wish. This is probably a better choice, as it seems to address both of your concerns (as well as not being in this package).

Thoughts?

**Comment by expipiplus1** [comment]
I think either way the Generic instance as it stands should go: it's too easy to do something like this: 

```haskell
> import Data.Vector.Sized
> import Data.Aeson
> instance FromJSON (Vector 3 Int)
> decode "[1,2,3,4]" :: Maybe (Vector 3 Int)
Just (Vector [1,2,3,4])
```

**Comment by kozross** [comment]
@expipiplus1 I agree. I think the 'morally correct' thing to do, in light of what @mstksg says, is 'no ``Generic`` instance as current, but no abstract/structural/'morally right' instance to replace it either'. Unless the ``NFData`` thing isn't a major blocker in your respective opinions.

The ``*JSON`` question is more fundamental, as it goes to things like 'what is the right way to represent a vector with a burned-in size in JSON', which I think are a bit outside our scope.

**Comment by mstksg** [comment]
Ultimately I think in the end the important thing to me is working with user expectations.  I'm not sure if either way of "keeping" a Generic instance will ever fully match user expectations in every situation...but I don't think we need a perfect instance to justify it being used.

In the end either way is justifiable and imperfect and usable to me :)  In an ideal world we'd be able to selectively hide typeclass instances to people can take advantage of whatever underlying representation they want to imagine things as for auto-deriving purposes.

However, some code that relies on Generic might change at this point, so we might be past that point depending on how much you want to preserve backwards compatibility :'(

**Comment by kozross** [comment]
@mstksg You make good points. At the same time, I believe that the PVP exists for a reason, and 'making breaking changes' is one of them. Not all APIs are made perfect - I literally released ``finitary`` a day ago and I'm _already_ seeing issues - and thus, I don't think it's the end of the world if we break something downstream. Additionally, you are probably _the_ biggest downstream consumer of ``vector-sized``: do you think that a significant chunk of stuff you've written would break on such a change?

For what it's worth, my ten cents is this: the whole point of libraries like ``vector-sized``, aside from enabling clever type-level stuff, is _safety_. By leaning on its public API, I gain a degree of confidence that stuff won't be dangerous, and won't expose anything that could be. While the Law of Leaky Abstractions is a thing, and all abstractions _do_ leak, at the very least, these leaks should be:

* Deliberate; and
* Well-documented

I agree that either 'get rid of ``Generic`` altogether' and 'keep it as-is' are defensible from different points of view. However, I believe that as current, the ``Generic`` instance is incredibly leaky (as @expipiplus1 so clearly showed), and thus, if we are to retain it, we need to make sure our documentation _practically yells_ "Hey! This is a possible abstraction leak!". It doesn't at this stage (or at least I don't recall it doing such). Furthermore, as @expipiplus1 showed, this can leak via stuff we don't even control.

At the very least, if we are to keep it, I advise (and volunteer for) documentation changes to indicate its leakiness. However, I believe the _principled_ thing to do is to remove it altogether. If we're concerned about important derivations (such as ``NFData``), we can provide them ourselves (and probably should in any case).

**Comment by expipiplus1** [comment]
As a small addition to the example above, as `vector-sized` uses the "unsafe" operations from `vector` it's possible to index out of bounds without mentioning "unsafe": 

```haskell
(`index'` (Proxy @2)) <$> (decode "[1]" :: Maybe (Vector 3 Int))
```

I think we should get rid of the `Generic` instance. I'm not too fussed about breaking backwards compatibility here; any user code which uses the generic instance shouldn't be too hard to write 
(more honestly) using the constructor exposed in the Internal module here.

**Comment by expipiplus1** [comment]
> The `*JSON` question is more fundamental, as it goes to things like 'what is the right way to represent a vector with a burned-in size in JSON', which I think are a bit outside our scope.

Just as an aside, aeson represents tuples using JSON Lists: 

```
*Data.Aeson> encode (1,2,3)
"[1,2,3]"
*Data.Aeson> encode (1,2,"hello")
"[1,2,\"hello\"]"
```

**Comment by kozross** [comment]
@expipiplus1 Agreed. Are you happy to make the change?

**Comment by expipiplus1** [comment]
Released as v1.4.0.0
